explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
   }
 }
 
+// A abstracted view of a pattern or constant value (which we view as a
+// particular kind of pattern).
+class AbstractPattern {
+ public:
+  enum Kind {
+    // A pattern that matches anything.
+    Wildcard,
+    // A pattern that matches a compound value with sub-patterns to match
+    // elements. A compound value is modeled as a discriminator name applied to
+    // a sequence of nested values: the alternative `Optional.Element(E)` has
+    // discriminator `Element` and nested value `E`, and the tuple `(A,B,C)`
+    // has an empty discriminator and nested values `A`, `B`, and `C`.
+    Compound,
+    // A pattern that matches a particular primitive value.
+    Primitive
+  };
+
+  AbstractPattern(Nonnull<const Pattern*> pattern) { Set(pattern); }
```

Comment:
```suggestion
  explicit AbstractPattern(Nonnull<const Pattern*> pattern) { Set(pattern); }
```

Reply:
It's important for the ergonomics elsewhere that this conversion is implicit, eg filling a vector of `AbstractPattern`s from a range of `const Pattern*`s.

Reply:
Can you then add `// NOLINTNEXTLINE(google-explicit-constructor)` and an explanation?

Reply:
Done.